src/Routes/Abstract_Route.php

```diff
@@ -0,0 +1,356 @@
+<?php // phpcs:disable WordPress.Files.FileName.NotHyphenatedLowercase
+/**
+ * Route is the base class for Routes. At the least a Route must
+ * declare,
+ *
+ * 1. Name
+ * 2. Pattern
+ * 3. Matches
+ * 4. Template
+ *
+ * It can also hook into the route's lifecyle to alter the WP & WP_Query
+ * objects by overriding parse and before methods.
+ *
+ * Usage: To create a new route /things/<thing>/ at ThingRoute,
+ *
+ * class ThingRoute extends Route {
+ *
+ *     public function get_name() {
+ *         return 'things';
+ *     }
+ *
+ *     public function get_pattern() {
+ *         return 'things/([^/]+)/?&';
+ *     }
+ *
+ *     public function get_matches() {
+ *         return [ 'thing' => '$matches[1]' ];
+ *     }
+ *
+ *     public function get_template() {
+ *         return 'thing.php';
+ *     }
+ *
+ * }
+ *
+ * @package Tribe\Project\Routes
+ */
+
+declare( strict_types=1 );
+
+namespace Tribe\Libs\Routes;
+
+/**
+ * Base class for routes.
+ */
+abstract class Abstract_Route {
+	/**
+	 * Registers any hooks for the route.
+	 *
+	 * @return void
+	 */
+	abstract public function register() : void;
```

Comment:
As mentioned in the other PR, we should not register any hooks in our instances, so probably won't need this.